src/main/kotlin/org/jetbrains/research/tasktracker/tracking/activity/ActivityTracker.kt

```diff
@@ -0,0 +1,123 @@
+import com.intellij.execution.process.BaseProcessHandler
+import com.intellij.execution.process.ProcessHandler
+import com.intellij.execution.runners.ExecutionEnvironment
+import com.intellij.ide.IdeEventQueue
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.actionSystem.ActionManager
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.ex.AnActionListener
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Disposer
+import com.intellij.util.messages.MessageBusConnection
+import com.intellij.util.messages.Topic
+import org.jetbrains.research.tasktracker.tracking.activity.actions.MovingActions
+import org.jetbrains.research.tasktracker.tracking.activity.actions.TypingActions
+import org.jetbrains.research.tasktracker.tracking.logger.ActivityLogger
+import java.awt.AWTEvent
+import java.awt.event.KeyEvent
+import java.util.concurrent.atomic.AtomicInteger
+
+class ActivityTracker(project: Project) {
+    private val activityLogger: ActivityLogger = ActivityLogger(project)
+    private val messageBusConnections: MutableList<MessageBusConnection> = mutableListOf()
+    private var trackingDisposable: Disposable? = null
+    private val currentId: AtomicInteger = AtomicInteger(1)
+
+    // TODO: add config to select activities to track
+    fun startTracking() {
+        trackingDisposable = Disposer.newDisposable()
+        listenActions()
+        listenKeyboard()
+        listenExecution()
+    }
+
+    fun stopTracking() {
+        messageBusConnections.forEach { it.disconnect() }
+        trackingDisposable?.let { Disposer.dispose(it) }
+    }
+
+    private fun listenActions(listenShortcut: Boolean = true) {
+        val actionListener = object : AnActionListener {
+            override fun beforeActionPerformed(anAction: AnAction, event: AnActionEvent) {
+                val actionId = ActionManager.getInstance().getId(anAction)
+                    ?: anAction.javaClass.name
+                val id = currentId.getAndIncrement()
+                if (listenShortcut) {
+                    actionId.extractShortcut(event, id)
+                }
+                activityLogger.log(Type.Action, actionId, id)
+            }
+        }
+        actionListener.connect(AnActionListener.TOPIC)
+    }
+
+    private fun String.extractShortcut(event: AnActionEvent, id: Int) {
+        if (!movingActions.contains(this) && !typingActions.contains(this)) {
+            val input = event.inputEvent
+            if (input is KeyEvent) {
+                activityLogger.log(Type.Shortcut, eventToShortcutInfo(input), id)
+            }
+        }
+    }
+
+    private fun eventToShortcutInfo(event: KeyEvent): String {
+        val keys = Key.values().mapNotNull { it.check(event) }.map { it.name }.plus(event.keyCode.toString())
+        return keys.joinToString(" ")
```

Comment:
You can merge these two operations and rewrite this function as a function expression